WIP on port to Swift3 Dispatch API

dgrove-oss opened the pull request:
Swift code compiles against current master branch of Dispatch,
with Swift3 wrapping overlay, but doesn't link
(need to port dispatch_get_main_queue_eventfd_4CF
and dispatch_main_queue_drain_4CF from experimental/foundation
to master).

Note: this pull request is not ready to merge; posting just for documentation and to have it available
when we are ready to merge.

dgrove-oss commented on Foundation/NSOperation.swift:
```diff
@@ -406,27 +405,27 @@ public class OperationQueue: NSObject {
                 } else {
                     self._runOperation()
                 }
                 if let group = waitGroup {
```
The DispatchWorkItem API is going to change next week per discussion in https://github.com/apple/swift-corelibs-libdispatch/pull/100.

parkera commented:
This all looks good to me - should I merge it or are we waiting on dispatch changes?

dgrove-oss commented:
I think we need to at least wait on the dispatch changes being merged.  After that, I'm not sure what the next step is.  After this is merged, the code won't build using the experimental/foundation branch of dispatch.  So there may need to be some coordination with the CI environment, etc (maybe not a good thing to merge on a Friday afternoon ;) )  @seabaylea any thoughts on staging?

dgrove-oss commented:
argh.  I blew a rebase operation.  Sorry. Will try to recover;  definitely don't merge this yet.

dgrove-oss commented:
Rebased with master as of last night and updated for the final cf runloop dispatch API that was merged yesterday to master in dispatch.   You can merge this whenever you are ready to switch to the master branch of dispatch.

parkera commented:
Great, I'll coordinate with @shahmishal and we'll get this going.

shahmishal commented:
@swift-ci please test

parkera commented:
Let's do it!
